### ISSUE TITLE
[mini.test] Ability to ignore screenshot attributes

### ISSUE THREAD

**Issue by stasjok** [echasnovski/mini.nvim#1300]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.test

### Description

Screenshots give the ability to compare what is displayed on the screen and how it is displayed. But sometimes I only care of _what_ is displayed, not _how_ it's displayed. The latter can change from the external changes like colorscheme or something. I think in some cases an additional option to ignore screenshot attributes can be useful.

**Comment by echasnovski** [comment]
Thanks for the suggestion!

The use case seems more or less valid, but an actual implementation would require a bit messy source code changes. Mostly because for completeness it would also require to allow users to ignore text in favor of comparing only attributes, which is not ideal for my taste.

In general, I think that validating screenshots in full (both text and attributes) is a good practice for a reliable test. Here is what you can do to make comparing only text less painful:
- Manually set custom color scheme. The one over which you have full control. This is [what is done in 'mini.nvim'](https://github.com/echasnovski/mini.nvim/blob/8d94201fece82559881362da80063a4842b6c4c4/scripts/minimal_init.lua#L5-L13).
- If troubles come from only a certain known lines (which don't contain tested text), then you can use [`opts.ignore_lines`](https://github.com/echasnovski/mini.nvim/blob/8d94201fece82559881362da80063a4842b6c4c4/doc/mini-test.txt#L584). This is commonly useful to ignore command line text/attributes or lines that contain text varying on testing instance (like full paths).
- Resort to testing buffer text explicitly via comparing `child.api.nvim_buf_get_lines(0, 0, -1, false)` to a reference array.

Closing as not planned.

**Comment by echasnovski** [comment]
Thanks for adopting 'mini.test' usage!

Did you try setting custom color scheme or is it the case of when Nightly has some different attributes in Command-line area? In the latter cases I usually resort to only do screenshot testing on Nightly+ (or sometimes *only* on stable release) which is also a bit more future proof.

In the hindsight, the `ignore_lines` should have been separate `ingore_text_lines` and `ignore_attr_lines`, but its original use case (ignoring screenshot parts which contain full path that depends on the test runner) really needs setting both.

As an appreciation for using 'mini.test' in such a vital plugin, let's reopen this so that I'll think about it more after next 'mini.nvim' stable release.

**Comment by ibhagwan** [comment]
> Did you try setting custom color scheme or is it the case of when Nightly has some different attributes in Command-line area? In the latter cases I usually resort to only do screenshot testing on Nightly+ (or sometimes only on stable release) which is also a bit more future proof.

I saw the suggestion but didn’t try that yet, I felt avoiding the attrs was more prudent in that specific test case (basic ui layout and command result), there are other places I do need the attr check on all platforms/versions and plan to do so for these (I just started to add ui screenshot tests).

> In the hindsight, the ignore_lines should have been separate ingore_text_lines and ignore_attr_lines, but its original use case (ignoring screenshot parts which contain full path that depends on the test runner) really needs setting both.

Correct, paths are more problematic on screenshots due to the different dir structures of the CI env, also helps me to test windows using dir separator without forcing path normalization to use backslashes.

> As an appreciation for using 'mini.test' in such a vital plugin, let's reopen this so that I'll think about it more after next 'mini.nvim' stable release.

Ty @echasnovski! Absolutely no rush on this as my workaround code works great (ty @MagicDuck).

**Comment by ibhagwan** [comment]
A few other things I forgot to mention:

- The current workaround sets the attr lines to empty lines resulting in a screenshot file half empty at the bottom, it would be nice to have the screenshot file reduce to just the visible lines, will also make the expect message nicer to distinguish between reference and observed.


While reading TESTING.md I came across:

> Sometimes hanging process will occur: it stops executing without any output. Most of the time it is because Neovim process is "blocked", i.e. it waits for user input and won't return from other call. Common causes are active hit-enter-prompt (solution: increase prompt height to a bigger value) or Operator-pending mode (solution: exit it). To mitigate this experience, most helper methods will throw an error if they can deduce that immediate execution will lead to hanging state.

Perhaps it’s query mentioning that this can be also overcome by using the new `:help mopt` in the nightly:
```vimdoc
'messagesopt' 'mopt'	string	(default "hit-enter,history:500")
			global
	Option settings for outputting messages.  It can consist of the
	following items.  Items must be separated by a comma.

	hit-enter	Use a |hit-enter| prompt when the message is longer than
			'cmdheight' size.
 ```

**Comment by echasnovski** [comment]
So after a lot of back and forth I decided to go ahead and replace `ignore_lines` with a more flexible `ignore_text` and `ignore_attr`. Both of new options can be either an array of number for lines to ignore text/attr *or* a boolean indicating whether to ignore the whole part (for example, `ignore_attr = true` ignores any highlighting when comparing screenshots).

The downside is that leaving `ignore_lines` felt redundant, so it is now soft deprecated. It will work at least until the next release (but with annoying warning) and will be removed some time afterwards. @ibhagwan, if anybody will complain about that, I'll tell them that it's your fault :)

**Comment by ibhagwan** [comment]
> So after a lot of back and forth I decided to go ahead and replace `ignore_lines` with a more flexible `ignore_text` and `ignore_attr`. Both of new options can be either an array of number for lines to ignore text/attr _or_ a boolean indicating whether to ignore the whole part (for example, `ignore_attr = true` ignores any highlighting when comparing screenshots).
> 
> The downside is that leaving `ignore_lines` felt redundant, so it is now soft deprecated. It will work at least until the next release (but with annoying warning) and will be removed some time afterwards. [@ibhagwan](https://github.com/ibhagwan), if anybody will complain about that, I'll tell them that it's your fault :)

Tysm @echasnovski, no problem I’ll take the blame :)

@phanen, we probably have to change this too: https://github.com/ibhagwan/fzf-lua/commit/ea6d8ce0842c9a0d16587fcaaa75efe64e8c5695.

**Comment by MagicDuck** [comment]
this is exciting! might be time for me to clean up all the hacks in grug-far.nvim tests 😅

**Comment by MagicDuck** [comment]
hmm, tried to use it today, but it looks like the change is just about ignoring the attrs, it doesn't also remove them when creating the screenshot or when reporting Observed/Stored. I think I'll stick to what I have for now.

**Comment by echasnovski** [comment]
> hmm, tried to use it today, but it looks like the change is just about ignoring the attrs, it doesn't also remove them when creating the screenshot or when reporting Observed/Stored. I think I'll stick to what I have for now.

Yes, of course. Create/get screenshot and compare to reference screenshot are two different operations. The screenshot format is fixed (at least until there is a good way to get exact cell *colors* to store as "attribute map legend" of sorts) and includes both text and attribute parts. How the reference screenshot gets compared to the target one is subject of configuration here.

**Comment by MagicDuck** [comment]
thanks for the clarification! was my misunderstanding
